pep-0544.txt

```diff
@@ -0,0 +1,967 @@
+before it could be instantiated.
+
+
+Merging and extending protocols
+-------------------------------
+
+The general philosophy is that protocols are mostly like regular ABCs,
+but a static type checker will handle them specially. Subclassing a protocol
+class would not turn the subclass into a protocol unless it also has
+``typing.Protocol`` as an explicit base class. Without this base, the class
+is "downgraded" to a regular ABC that cannot be used with structural
+subtyping.
+
+A subprotocol can be defined by having *both* one or more protocols as
+immediate base classes and also having ``typing.Protocol`` as an immediate
+base class::
+
+  from typing import Sized, Protocol
+
+  class SizedAndCloseable(Sized, Protocol):
+      def close(self) -> None:
+          ...
+
+Now the protocol ``SizedAndCloseable`` is a protocol with two methods,
+``__len__`` and ``close``. If one omits ``Protocol`` in the base class list,
+this would be a regular (non-protocol) class that must implement ``Sized``.
+If ``Protocol`` is included in the base class list, all the other base classes
+must be protocols. A protocol can't extend a regular class.
+
+Alternatively, one can implement ``SizedAndCloseable`` like this, assuming
+the existence of ``SupportsClose`` from the example in `definition`_ section::
+
+  from typing import Sized
+
+  class SupportsClose(...): ...  # Like above
+
+  class SizedAndCloseable(Sized, SupportsClose, Protocol):
+      pass
+
+The two definitions of ``SizedAndClosable`` are equivalent.
+Subclass relationships between protocols are not meaningful when
+considering subtyping, since structural compatibility is
+the criterion, not the MRO.
+
+Note that rules around explicit subclassing are different from regular ABCs,
+where abstractness is simply defined by having at least one abstract method
+being unimplemented. Protocol classes must be marked *explicitly*.
+
+
+Generic and recursive protocols
+-------------------------------
+
+Generic protocols are important. For example, ``SupportsAbs``, ``Iterable``
+and ``Iterator`` are generic protocols. They are defined similar to normal
+non-protocol generic types::
+
+  T = TypeVar('T', covariant=True)
+
+  class Iterable(Protocol[T]):
+      @abstractmethod
+      def __iter__(self) -> Iterator[T]:
+          ...
+
+Note that ``Protocol[T, S, ...]`` is allowed as a shorthand for
+``Protocol, Generic[T, S, ...]``.
+
+Recursive protocols are also supported. Forward references to the protocol
+class names can be given as strings as specified by PEP 484. Recursive
+protocols will be useful for representing self-referential data structures
+like trees in an abstract fashion::
+
+  class Traversable(Protocol):
+      leaves: Iterable['Traversable']
+
+
+Using Protocols
+===============
+
+Subtyping relationships with other types
+----------------------------------------
+
+Protocols cannot be instantiated, so there are no values with
```

Comment:
That sounds a little weird -- certainly an instance of a class that implements a protocol is a value of that protocol type? Maybe all you need is to add "exactly" somewhere to the sentence.